File: internal/ocr/parse.go
package ocr

import (
	"bytes"
	"encoding/json"
	"fmt"
	"strconv"
	"strings"
	"time"
	"unicode"

	"github.com/shopspring/decimal"
)

func Parse(raw []byte) (Bill, error) {
	return parseBill(raw)
}

func parseBill(raw []byte) (Bill, error) {
	payload, err := extractJSON(raw)
	if err != nil {
		return Bill{}, err
	}
	var bill Bill
	if err := json.Unmarshal(payload, &bill); err != nil {
		return Bill{}, fmt.Errorf("model returned invalid JSON")
	}
	bill.Notes = strings.TrimSpace(bill.Notes)
	bill.StoryName = strings.TrimSpace(bill.StoryName)
	bill.ExternalID = strings.TrimSpace(bill.ExternalID)
	bill.StreetName = stripStreetPrefix(bill.StreetName)
	bill.BuildingNumber = strings.TrimSpace(bill.BuildingNumber)
	bill.ApartmentNumber = strings.TrimSpace(bill.ApartmentNumber)
	bill.PostalCode = normalizePostal(bill.PostalCode)
	bill.City = strings.TrimSpace(bill.City)
	bill.BoughtOn, bill.BoughtAt = splitDateAndTime(bill.BoughtOn, bill.BoughtAt)
	bill.BoughtOn = normalizeDate(bill.BoughtOn)
	bill.BoughtAt = normalizeTime(bill.BoughtAt)
	for i := range bill.Lines {
		normalizeLine(&bill.Lines[i])
		fixWeighedKg(&bill.Lines[i])
		fillMissingAmount(&bill.Lines[i])
	}
	return bill, nil
}

func normalizeLine(line *Line) {
	line.ReceiptName = strings.TrimSpace(line.ReceiptName)
	line.ProductName = strings.TrimSpace(line.ProductName)
	if line.ProductName == "" {
		line.ProductName = line.ReceiptName
	}
	line.UnitName = strings.TrimSpace(line.UnitName)
	line.SkipReason = strings.TrimSpace(line.SkipReason)

	unitPrice, vatFromPrice := peelVAT(line.UnitPrice)
	amount, vatFromAmount := peelVAT(line.Amount)
	count, vatFromCount := peelVAT(line.PackageCount)
	line.UnitPrice = normalizeNumber(unitPrice)
	line.Amount = normalizeNumber(amount)
	line.PackageCount = normalizeNumber(count)
	line.PackageSize = normalizeNumber(line.PackageSize)
	line.Quantity = normalizeNumber(line.Quantity)
	line.Discount = normalizeDiscount(line.Discount)
	line.VatType = normalizeVAT(line.VatType, vatFromPrice, vatFromAmount, vatFromCount)
}

func extractJSON(raw []byte) ([]byte, error) {
	s := strings.TrimSpace(string(raw))
	if s == "" {
		return nil, fmt.Errorf("empty model response")
	}
	if i := strings.Index(s, "```"); i >= 0 {
		s = s[i+3:]
		s = strings.TrimSpace(strings.TrimPrefix(s, "json"))
		if j := strings.Index(s, "```"); j >= 0 {
			s = s[:j]
		}
		s = strings.TrimSpace(s)
	}
	start := strings.IndexByte(s, '{')
	end := strings.LastIndexByte(s, '}')
	if start < 0 || end < start {
		return nil, fmt.Errorf("model response was not JSON")
	}
	return []byte(s[start : end+1]), nil
}

func splitDateAndTime(date, clock string) (string, string) {
	date = strings.TrimSpace(date)
	clock = strings.TrimSpace(clock)
	date = strings.ReplaceAll(date, "\u00a0", " ")
	date = strings.ReplaceAll(date, "T", " ")
	parts := strings.Fields(date)
	if len(parts) == 0 {
		return date, clock
	}
	date = parts[0]
	if clock == "" && len(parts) >= 2 {
		clock = parts[1]
	}
	return date, clock
}

func normalizeDate(s string) string {
	s = strings.TrimSpace(s)
	if s == "" {
		return ""
	}
	s = strings.ReplaceAll(s, "\u00a0", " ")
	if t, err := time.Parse("2006-01-02", s); err == nil {
		return t.Format("2006-01-02")
	}
	for _, layout := range []string{
		"02.01.2006",
		"2.01.2006",
		"02.1.2006",
		"2.1.2006",
		"02/01/2006",
		"2/1/2006",
		"02-01-2006",
		"2006/01/02",
		"2006.01.02",
	} {
		if t, err := time.Parse(layout, s); err == nil {
			return t.Format("2006-01-02")
		}
	}
	return s
}

func normalizeTime(s string) string {
	s = strings.TrimSpace(s)
	if s == "" {
		return ""
	}
	s = strings.ReplaceAll(s, "\u00a0", " ")
	s = strings.ReplaceAll(s, ".", ":")
	if i := strings.IndexByte(s, ' '); i >= 0 {
		s = s[:i]
	}
	for _, layout := range []string{"15:04:05", "15:04", "15:04:05Z07:00"} {
		if t, err := time.Parse(layout, s); err == nil {
			return t.Format("15:04")
		}
	}
	return s
}

func normalizePostal(s string) string {
	s = strings.TrimSpace(s)
	if s == "" {
		return ""
	}
	var digits strings.Builder
	for _, r := range s {
		if unicode.IsDigit(r) {
			digits.WriteRune(r)
		}
	}
	d := digits.String()
	if len(d) == 5 {
		return d[:2] + "-" + d[2:]
	}
	return s
}

func stripStreetPrefix(s string) string {
	s = strings.TrimSpace(s)
	if s == "" {
		return ""
	}
	lower := strings.ToLower(s)
	for _, p := range []string{"ulica ", "ul. ", "ul ", "aleje ", "aleja ", "al. ", "al ", "plac ", "pl. ", "pl "} {
		if strings.HasPrefix(lower, p) {
			return strings.TrimSpace(s[len(p):])
		}
	}
	return s
}

func peelVAT(s string) (num, vat string) {
	s = strings.TrimSpace(s)
	if s == "" {
		return "", ""
	}
	runes := []rune(s)
	last := unicode.ToUpper(runes[len(runes)-1])
	if last == 'A' || last == 'B' || last == 'C' {
		return strings.TrimSpace(string(runes[:len(runes)-1])), string(last)
	}
	return s, ""
}

func normalizeVAT(values ...string) string {
	for _, s := range values {
		s = strings.ToUpper(strings.TrimSpace(s))
		if s == "" {
			continue
		}
		s = strings.TrimRightFunc(s, func(r rune) bool {
			return unicode.IsDigit(r) || r == '%' || unicode.IsSpace(r)
		})
		s = strings.TrimSpace(s)
		if s == "A" || s == "B" || s == "C" {
			return s
		}
		if r := []rune(s); len(r) > 0 {
			last := unicode.ToUpper(r[len(r)-1])
			if last == 'A' || last == 'B' || last == 'C' {
				return string(last)
			}
		}
	}
	return ""
}

func normalizeDiscount(s string) string {
	s = strings.TrimSpace(s)
	s = strings.TrimPrefix(s, "−")
	s = strings.TrimPrefix(s, "–")
	s = strings.TrimPrefix(s, "-")
	s = normalizeNumber(s)
	if s == "" {
		return ""
	}
	d, err := decimal.NewFromString(s)
	if err != nil {
		return s
	}
	return d.Abs().StringFixed(2)
}

func normalizeNumber(s string) string {
	s = strings.TrimSpace(s)
	s = strings.ReplaceAll(s, "\u00a0", "")
	s = strings.ReplaceAll(s, " ", "")
	s = strings.Map(func(r rune) rune {
		if unicode.IsLetter(r) || r == '%' {
			return -1
		}
		return r
	}, s)
	s = strings.ReplaceAll(s, ",", ".")
	if strings.Count(s, ".") > 1 {
		parts := strings.Split(s, ".")
		s = strings.Join(parts[:len(parts)-1], "") + "." + parts[len(parts)-1]
	}
	return s
}

// fillMissingAmount sets amount from printed unit price, qty, and rabat
// when the till did not print a final line total. Never overwrites a
// printed amount.
func fillMissingAmount(line *Line) {
	if line.Skip || strings.TrimSpace(line.Amount) != "" {
		return
	}
	price, err := decimal.NewFromString(line.UnitPrice)
	if err != nil || price.IsZero() {
		return
	}
	count, err := decimal.NewFromString(line.PackageCount)
	if err != nil || count.IsZero() {
		return
	}
	discount := decimal.Zero
	if line.Discount != "" {
		d, err := decimal.NewFromString(line.Discount)
		if err == nil {
			discount = d.Abs()
		}
	}
	line.Amount = count.Mul(price).Sub(discount).Round(2).StringFixed(2)
}

// fixWeighedKg maps "1 pack of 1.450 kg" (old OCR rule) to "1.450 × 1 kg"
// when the size looks like a till scale reading (three decimal places).
func fixWeighedKg(line *Line) {
	if !strings.EqualFold(line.UnitName, "kg") {
		return
	}
	if !isOne(line.PackageCount) {
		return
	}
	if !looksLikeScaleKg(line.PackageSize) {
		return
	}
	line.PackageCount = line.PackageSize
	line.PackageSize = "1"
}

func isOne(s string) bool {
	switch s {
	case "1", "1.0", "1.00", "1.000":
		return true
	}
	return false
}

func looksLikeScaleKg(s string) bool {
	i := strings.IndexByte(s, '.')
	if i < 0 {
		return false
	}
	frac := s[i+1:]
	if len(frac) < 3 {
		return false
	}
	return strings.TrimRight(frac, "0") != ""
}

type flexInt int64

func (n *flexInt) UnmarshalJSON(b []byte) error {
	b = bytes.TrimSpace(b)
	if len(b) == 0 || bytes.Equal(b, []byte("null")) {
		*n = 0
		return nil
	}
	if b[0] == '"' {
		var s string
		if err := json.Unmarshal(b, &s); err != nil {
			return err
		}
		s = strings.TrimSpace(s)
		if s == "" {
			*n = 0
			return nil
		}
		v, err := strconv.ParseInt(s, 10, 64)
		if err != nil {
			*n = 0
			return nil
		}
		*n = flexInt(v)
		return nil
	}
	var v float64
	if err := json.Unmarshal(b, &v); err != nil {
		*n = 0
		return nil
	}
	*n = flexInt(v)
	return nil
}

type flexNum string

func (n *flexNum) UnmarshalJSON(b []byte) error {
	b = bytes.TrimSpace(b)
	if len(b) == 0 || bytes.Equal(b, []byte("null")) {
		*n = ""
		return nil
	}
	if b[0] == '"' {
		var s string
		if err := json.Unmarshal(b, &s); err != nil {
			return err
		}
		*n = flexNum(s)
		return nil
	}
	*n = flexNum(bytes.TrimSpace(b))
	return nil
}

type rawBill struct {
	BoughtOn        string    `json:"bought_on"`
	BoughtAt        string    `json:"bought_at"`
	Notes           string    `json:"notes"`
	NotABill        bool      `json:"not_a_bill"`
	StoryID         flexInt   `json:"company_id"`
	StoryName       string    `json:"company_name"`
	ExternalID      string    `json:"external_id"`
	StreetName      string    `json:"street_name"`
	BuildingNumber  string    `json:"building_number"`
	ApartmentNumber string    `json:"apartment_number"`
	PostalCode      string    `json:"postal_code"`
	City            string    `json:"city"`
	Lines           []rawLine `json:"lines"`
}

type rawLine struct {
	ReceiptName  string  `json:"receipt_name"`
	ProductName  string  `json:"product_name"`
	ProductID    flexInt `json:"product_id"`
	UnitID       flexInt `json:"unit_id"`
	UnitName     string  `json:"unit_name"`
	VatType      string  `json:"vat_type"`
	PackageCount flexNum `json:"package_count"`
	PackageSize  flexNum `json:"package_size"`
	Quantity     flexNum `json:"quantity"`
	UnitPrice    flexNum `json:"unit_price"`
	Discount     flexNum `json:"discount"`
	Amount       flexNum `json:"amount"`
	Skip         bool    `json:"skip"`
	SkipReason   string  `json:"skip_reason"`
}

func (b *Bill) UnmarshalJSON(data []byte) error {
	var raw rawBill
	if err := json.Unmarshal(data, &raw); err != nil {
		return err
	}
	b.BoughtOn = raw.BoughtOn
	b.BoughtAt = raw.BoughtAt
	b.Notes = raw.Notes
	b.NotABill = raw.NotABill
	b.StoryID = int64(raw.StoryID)
	b.StoryName = raw.StoryName
	b.ExternalID = raw.ExternalID
	b.StreetName = raw.StreetName
	b.BuildingNumber = raw.BuildingNumber
	b.ApartmentNumber = raw.ApartmentNumber
	b.PostalCode = raw.PostalCode
	b.City = raw.City
	b.Lines = make([]Line, len(raw.Lines))
	for i, line := range raw.Lines {
		b.Lines[i] = Line{
			ReceiptName:  line.ReceiptName,
			ProductName:  line.ProductName,
			ProductID:    int64(line.ProductID),
			UnitID:       int64(line.UnitID),
			UnitName:     line.UnitName,
			VatType:      line.VatType,
			PackageCount: string(line.PackageCount),
			PackageSize:  string(line.PackageSize),
			Quantity:     string(line.Quantity),
			UnitPrice:    string(line.UnitPrice),
			Discount:     string(line.Discount),
			Amount:       string(line.Amount),
			Skip:         line.Skip,
			SkipReason:   line.SkipReason,
		}
	}
	return nil
}

File: src/ocr/parse.ts
import Decimal from 'decimal.js';
import { emptyBill, emptyLine, type Bill, type Line } from './types';

export function parse(raw: Buffer | string): Bill {
  return parseBill(raw);
}

export function parseBill(raw: Buffer | string): Bill {
  const payload = extractJSON(typeof raw === 'string' ? raw : raw.toString('utf8'));
  let parsed: unknown;
  try {
    parsed = JSON.parse(payload);
  } catch {
    throw new Error('model returned invalid JSON');
  }
  const bill = unmarshalBill(parsed);
  bill.Notes = bill.Notes.trim();
  bill.StoryName = bill.StoryName.trim();
  bill.ExternalID = bill.ExternalID.trim();
  bill.StreetName = stripStreetPrefix(bill.StreetName);
  bill.BuildingNumber = bill.BuildingNumber.trim();
  bill.ApartmentNumber = bill.ApartmentNumber.trim();
  bill.PostalCode = normalizePostal(bill.PostalCode);
  bill.City = bill.City.trim();
  const split = splitDateAndTime(bill.BoughtOn, bill.BoughtAt);
  bill.BoughtOn = normalizeDate(split.date);
  bill.BoughtAt = normalizeTime(split.clock);
  for (const line of bill.Lines) {
    normalizeLine(line);
    inferUnitFromSize(line);
    fixWeighedKg(line);
    coalesceQuantity(line);
    inferUnitFromQuantity(line);
    fillMissingAmount(line);
  }
  return bill;
}

function unmarshalBill(parsed: unknown): Bill {
  if (parsed === null || typeof parsed !== 'object' || Array.isArray(parsed)) {
    throw new Error('model returned invalid JSON');
  }
  const raw = parsed as Record<string, unknown>;
  const bill = emptyBill();
  bill.BoughtOn = asStringField(raw.bought_on);
  bill.BoughtAt = asStringField(raw.bought_at);
  bill.Notes = asStringField(raw.notes);
  bill.NotABill = asBoolField(raw.not_a_bill);
  bill.StoryID = flexInt(raw.company_id);
  bill.StoryName = asStringField(raw.company_name);
  bill.ExternalID = asStringField(raw.external_id);
  bill.StreetName = asStringField(raw.street_name);
  bill.BuildingNumber = asStringField(raw.building_number);
  bill.ApartmentNumber = asStringField(raw.apartment_number);
  bill.PostalCode = asStringField(raw.postal_code);
  bill.City = asStringField(raw.city);
  if (raw.lines == null) {
    bill.Lines = [];
  } else if (!Array.isArray(raw.lines)) {
    throw new Error('model returned invalid JSON');
  } else {
    bill.Lines = raw.lines.map(unmarshalLine);
  }
  return bill;
}

function unmarshalLine(item: unknown): Line {
  if (item === null || typeof item !== 'object' || Array.isArray(item)) {
    throw new Error('model returned invalid JSON');
  }
  const raw = item as Record<string, unknown>;
  const line = emptyLine();
  line.ReceiptName = asStringField(raw.receipt_name);
  line.ProductName = asStringField(raw.product_name);
  line.ProductID = flexInt(raw.product_id);
  line.UnitID = flexInt(raw.unit_id);
  line.UnitName = asStringField(raw.unit_name);
  line.VatType = asStringField(raw.vat_type);
  line.PackageCount = flexNum(raw.package_count);
  line.PackageSize = flexNum(raw.package_size);
  line.Quantity = flexNum(raw.quantity);
  line.UnitPrice = flexNum(raw.unit_price);
  line.Discount = flexNum(raw.discount);
  line.Amount = flexNum(raw.amount);
  line.Skip = asBoolField(raw.skip);
  line.SkipReason = asStringField(raw.skip_reason);
  return line;
}

function asStringField(v: unknown): string {
  if (v == null) return '';
  if (typeof v === 'string') return v;
  throw new Error('model returned invalid JSON');
}

function asBoolField(v: unknown): boolean {
  if (v == null) return false;
  if (typeof v === 'boolean') return v;
  throw new Error('model returned invalid JSON');
}

function flexInt(v: unknown): number {
  if (v == null) return 0;
  if (typeof v === 'string') {
    const s = v.trim();
    if (s === '') return 0;
    const n = Number.parseInt(s, 10);
    return Number.isFinite(n) ? n : 0;
  }
  if (typeof v === 'number' && Number.isFinite(v)) return Math.trunc(v);
  if (typeof v === 'boolean') return 0;
  return 0;
}

function flexNum(v: unknown): string {
  if (v == null) return '';
  if (typeof v === 'string') return v;
  if (typeof v === 'number' && Number.isFinite(v)) {
    if (Number.isInteger(v)) return String(v);
    return String(v);
  }
  if (typeof v === 'boolean') return '';
  return '';
}

function normalizeLine(line: Line): void {
  line.ReceiptName = line.ReceiptName.trim();
  line.ProductName = line.ProductName.trim();
  if (line.ProductName === '') line.ProductName = line.ReceiptName;
  line.UnitName = normalizeUnitName(line.UnitName);
  line.SkipReason = line.SkipReason.trim();

  const [unitPrice, vatFromPrice] = peelVAT(line.UnitPrice);
  const [amount, vatFromAmount] = peelVAT(line.Amount);
  const [count, vatFromCount] = peelVAT(line.PackageCount);
  const [qty, vatFromQty] = peelVAT(line.Quantity);
  line.UnitPrice = normalizeNumber(unitPrice);
  line.Amount = normalizeNumber(amount);
  line.PackageCount = normalizeNumber(count);
  line.PackageSize = normalizeNumber(line.PackageSize);
  line.Quantity = normalizeNumber(qty);
  line.Discount = normalizeDiscount(line.Discount);
  line.VatType = normalizeVAT(line.VatType, vatFromPrice, vatFromAmount, vatFromCount, vatFromQty);
}

function extractJSON(raw: string): string {
  let s = raw.trim();
  if (s === '') throw new Error('empty model response');
  const fence = s.indexOf('```');
  if (fence >= 0) {
    s = s.slice(fence + 3);
    s = s.trim();
    if (s.startsWith('json')) s = s.slice(4);
    s = s.trim();
    const end = s.indexOf('```');
    if (end >= 0) s = s.slice(0, end);
    s = s.trim();
  }
  const start = s.indexOf('{');
  const last = s.lastIndexOf('}');
  if (start < 0 || last < start) throw new Error('model response was not JSON');
  return s.slice(start, last + 1);
}

function splitDateAndTime(date: string, clock: string): { date: string; clock: string } {
  date = date.trim();
  clock = clock.trim();
  date = date.replaceAll('\u00a0', ' ').replaceAll('T', ' ');
  const parts = date.split(/\s+/).filter((p) => p !== '');
  if (parts.length === 0) return { date, clock };
  date = parts[0]!;
  if (clock === '' && parts.length >= 2) clock = parts[1]!;
  return { date, clock };
}

export function normalizeDate(s: string): string {
  s = s.trim();
  if (s === '') return '';
  s = s.replaceAll('\u00a0', ' ');
  const iso = parseYMD(s, '-', true);
  if (iso) return iso;
  const layouts: Array<() => string | null> = [
    () => parseDMY(s, '.', 2, 2),
    () => parseDMY(s, '.', 1, 2),
    () => parseDMY(s, '.', 2, 1),
    () => parseDMY(s, '.', 1, 1),
    () => parseDMY(s, '/', 2, 2),
    () => parseDMY(s, '/', 1, 1),
    () => parseDMY(s, '-', 2, 2),
    () => parseYMD(s, '/', true),
    () => parseYMD(s, '.', true),
  ];
  for (const tryParse of layouts) {
    const got = tryParse();
    if (got) return got;
  }
  return s;
}

function parseYMD(s: string, sep: string, padded: boolean): string | null {
  const parts = s.split(sep);
  if (parts.length !== 3) return null;
  const [ys, ms, ds] = parts;
  if (!ys || !ms || !ds) return null;
  if (ys.length !== 4) return null;
  if (padded) {
    if (ms.length !== 2 || ds.length !== 2) return null;
  } else if (ms.length < 1 || ds.length < 1) {
    return null;
  }
  const y = Number.parseInt(ys, 10);
  const m = Number.parseInt(ms, 10);
  const d = Number.parseInt(ds, 10);
  return formatYMD(y, m, d);
}

function parseDMY(s: string, sep: string, dayDigits: number, monthDigits: number): string | null {
  const parts = s.split(sep);
  if (parts.length !== 3) return null;
  const [ds, ms, ys] = parts;
  if (!ds || !ms || !ys) return null;
  if (ys.length !== 4) return null;
  if (dayDigits === 2 && ds.length !== 2) return null;
  if (dayDigits === 1 && ds.length < 1) return null;
  if (monthDigits === 2 && ms.length !== 2) return null;
  if (monthDigits === 1 && ms.length < 1) return null;
  const d = Number.parseInt(ds, 10);
  const m = Number.parseInt(ms, 10);
  const y = Number.parseInt(ys, 10);
  return formatYMD(y, m, d);
}

function formatYMD(y: number, m: number, d: number): string | null {
  if (!Number.isFinite(y) || !Number.isFinite(m) || !Number.isFinite(d)) return null;
  if (m < 1 || m > 12 || d < 1 || d > 31) return null;
  const dt = new Date(Date.UTC(y, m - 1, d));
  if (dt.getUTCFullYear() !== y || dt.getUTCMonth() !== m - 1 || dt.getUTCDate() !== d) return null;
  return `${String(y).padStart(4, '0')}-${String(m).padStart(2, '0')}-${String(d).padStart(2, '0')}`;
}

export function normalizeTime(s: string): string {
  s = s.trim();
  if (s === '') return '';
  s = s.replaceAll('\u00a0', ' ').replaceAll('.', ':');
  const space = s.indexOf(' ');
  if (space >= 0) s = s.slice(0, space);
  const withZone = /^(\d{2}):(\d{2}):(\d{2})(?:Z|[+-]\d{2}:\d{2})$/i.exec(s);
  const hms = withZone ?? /^(\d{2}):(\d{2}):(\d{2})$/.exec(s);
  if (hms) {
    const hh = Number.parseInt(hms[1]!, 10);
    const mm = Number.parseInt(hms[2]!, 10);
    const ss = Number.parseInt(hms[3]!, 10);
    if (hh <= 23 && mm <= 59 && ss <= 59) {
      return `${hms[1]}:${hms[2]}`;
    }
    return s;
  }
  const hm = /^(\d{2}):(\d{2})$/.exec(s);
  if (hm) {
    const hh = Number.parseInt(hm[1]!, 10);
    const mm = Number.parseInt(hm[2]!, 10);
    if (hh <= 23 && mm <= 59) return `${hm[1]}:${hm[2]}`;
  }
  return s;
}

function normalizePostal(s: string): string {
  s = s.trim();
  if (s === '') return '';
  let digits = '';
  for (const r of s) {
    if (r >= '0' && r <= '9') digits += r;
  }
  if (digits.length === 5) return `${digits.slice(0, 2)}-${digits.slice(2)}`;
  return s;
}

export function stripStreetPrefix(s: string): string {
  s = s.trim();
  if (s === '') return '';
  const lower = s.toLowerCase();
  for (const p of ['ulica ', 'ul. ', 'ul ', 'aleje ', 'aleja ', 'al. ', 'al ', 'plac ', 'pl. ', 'pl ']) {
    if (lower.startsWith(p)) return s.slice(p.length).trim();
  }
  return s;
}

function peelVAT(s: string): [string, string] {
  s = s.trim();
  if (s === '') return ['', ''];
  const runes = [...s];
  const last = runes[runes.length - 1]!.toUpperCase();
  if (last === 'A' || last === 'B' || last === 'C') {
    return [runes.slice(0, -1).join('').trim(), last];
  }
  return [s, ''];
}

function normalizeVAT(...values: string[]): string {
  for (let s of values) {
    s = s.toUpperCase().trim();
    if (s === '') continue;
    s = s.replace(/[\d%\s]+$/u, '').trim();
    if (s === 'A' || s === 'B' || s === 'C') return s;
    const runes = [...s];
    if (runes.length > 0) {
      const last = runes[runes.length - 1]!.toUpperCase();
      if (last === 'A' || last === 'B' || last === 'C') return last;
    }
  }
  return '';
}

function normalizeDiscount(s: string): string {
  s = s.trim();
  if (s.startsWith('−')) s = s.slice(1);
  if (s.startsWith('–')) s = s.slice(1);
  if (s.startsWith('-')) s = s.slice(1);
  s = normalizeNumber(s);
  if (s === '') return '';
  try {
    const d = new Decimal(s);
    return d.abs().toFixed(2, Decimal.ROUND_HALF_EVEN);
  } catch {
    return s;
  }
}

export function normalizeNumber(s: string): string {
  s = s.trim();
  s = s.replaceAll('\u00a0', '').replaceAll(' ', '');
  s = [...s]
    .filter((r) => !/[\p{L}%]/u.test(r))
    .join('');
  s = s.replaceAll(',', '.');
  const dots = s.split('.');
  if (dots.length > 2) {
    s = dots.slice(0, -1).join('') + '.' + dots[dots.length - 1];
  }
  return s;
}

function coalesceQuantity(line: Line): void {
  if (line.Quantity.trim() !== '') return;
  const c = line.PackageCount.trim();
  if (c !== '') line.Quantity = c;
}

function inferUnitFromSize(line: Line): void {
  if (line.UnitName !== '') return;
  if (looksLikeScaleKg(line.PackageSize)) line.UnitName = 'kg';
}

function inferUnitFromQuantity(line: Line): void {
  if (line.UnitName !== '') return;
  if (looksLikeScaleKg(line.Quantity)) line.UnitName = 'kg';
}

function normalizeUnitName(s: string): string {
  s = s.toLowerCase().trim();
  if (s.endsWith('.')) s = s.slice(0, -1);
  switch (s) {
    case 'sztuka':
    case 'sztuk':
    case 'pcs':
    case 'pc':
    case 'piece':
    case 'szt':
      return 'szt';
    case 'opak':
    case 'opakowanie':
    case 'op':
    case 'pkt':
      return 'pkt';
    case 'kilogram':
    case 'kilogramy':
    case 'kilogramów':
    case 'kg':
      return 'kg';
    case 'gram':
    case 'gramy':
    case 'gramów':
    case 'gr':
    case 'g':
      return 'g';
    case 'litr':
    case 'litry':
    case 'litrów':
    case 'liter':
    case 'l':
      return 'l';
    case 'mililitr':
    case 'mililitry':
    case 'ml':
      return 'ml';
    default:
      return s;
  }
}

function fillMissingAmount(line: Line): void {
  if (line.Skip || line.Amount.trim() !== '') return;
  let price: Decimal;
  try {
    price = new Decimal(line.UnitPrice);
  } catch {
    return;
  }
  if (price.isZero()) return;
  let count: Decimal;
  try {
    count = new Decimal(line.Quantity);
  } catch {
    return;
  }
  if (count.isZero()) return;
  let discount = new Decimal(0);
  if (line.Discount !== '') {
    try {
      discount = new Decimal(line.Discount).abs();
    } catch {
      /* ignore */
    }
  }
  line.Amount = count.mul(price).sub(discount).toDecimalPlaces(2, Decimal.ROUND_HALF_EVEN).toFixed(2);
}

function fixWeighedKg(line: Line): void {
  if (line.UnitName.toLowerCase() !== 'kg') return;
  if (!isOne(line.PackageCount)) return;
  if (!looksLikeScaleKg(line.PackageSize)) return;
  line.PackageCount = line.PackageSize;
  line.PackageSize = '1';
}

function isOne(s: string): boolean {
  return s === '1' || s === '1.0' || s === '1.00' || s === '1.000';
}

function looksLikeScaleKg(s: string): boolean {
  const i = s.indexOf('.');
  if (i < 0) return false;
  const frac = s.slice(i + 1);
  if (frac.length < 3) return false;
  return frac.replace(/0+$/, '') !== '';
}
